lib/object.jsonnet

```diff
@@ -0,0 +1,99 @@
+
+  // keys returns all the non-hidden field keys of @obj in an array in key
+  // order.
+  keys(obj):: std.objectFields(obj),
+
+  // keysAll returns all the field keys (hidden and non-hidden) of @obj in an
+  // array in key order.
+  keysAll(obj):: std.objectFieldsAll(obj),
+
+  // keysHidden returns all the hidden field keys of @obj in an array in key
+  // order.
+  keysHidden(obj):: std.setDiff(object.keysAll(obj), object.keys(obj)),
+
+  // vals returns all the non-hidden field values of @obj in an arary in key order.
+  vals(obj):: __vals(obj, object.keys),
+
+  // valsAll returns all the field values (hidden and non-hidden) of @obj in an
+  // array in key order.
+  valsAll(obj):: __vals(obj, object.keysAll),
+
+  // valsHidden returns all the hidden field values of @obj in an array in key
+  // order.
+  valsHidden(obj):: __vals(obj, object.keysHidden),
+
+  // kv returns all the non-hidden fields of @obj as an array of `[key, value]`
+  // pairs in key order.
+  kv(obj):: __kv(obj, object.keys),
+
+  // kvAll returns all the fields (hidden and non-hidden) of @obj as an array
+  // of `[key, value]` pairs in key order.
+  kvAll(obj):: __kv(obj, object.keysAll),
+
+  // kvHidden returns all the hidden fields of @obj as an array of
+  // `[key, value]` pairs in key order.
+  kvHidden(obj):: __kv(obj, object.keysHidden),
+
+  local __vals(obj, selector) = [obj[k] for k in selector(obj)],
+  local __kv(obj, selector) = [[k, obj[k]] for k in selector(obj)],
+
+  // make returns an object from an array @arr of `[key, value]` pairs. The
+  // fields in the returned object are all visible (non-hidden).
+  make(arr):: { [kv[0]]: kv[1] for kv in arr },
+
+  // makeHidden returns an object from an array @arr of `[key, value]` pairs.
+  // The fields in the returned object are all hidden.
+  makeHidden(arr):: array.accumulate([{ [kv[0]]:: kv[1] } for kv in arr], {}),
+
+  // transform calls @fn(key, value) on each field of @obj passing it the key
+  // and the value of each field and using the result of @fn to build the
+  // result of transform. It can remove fields, rename fields or change the
+  // value of fields based on what the given function @fn returns.
+  //
+  // If @fn returns null, the field is removed from the result of transform.
+  // Otherwise @fn should return a `[key, value]` pair which is put into the
+  // result of transform as a field. The visibility of the field in the result
+  // object is the same as the visibility of the field passed to @fn.
+  transform(fn, obj)::
+    local kvToArgs(fn) = function(kv) fn(kv[0], kv[1]);
+    local visible = array.collapse(std.map(kvToArgs(fn), object.kv(obj)));
+    local hidden = array.collapse(std.map(kvToArgs(fn), object.kvHidden(obj)));
+    object.make(visible) + object.makeHidden(hidden),
+
+  // filter calls @fn on each field of @obj passing it the key and the value
+  // of each field and returns an object with the fields for which @fn
+  // returned true.
+  filter(fn, obj)::
```

Comment:
I was about to ask whether should also be a filer for arrays, then I checked and saw: `std.filter(func, arr)` on 
https://jsonnet.org/ref/stdlib.html (docs need anchors imo). I wonder if a references would make sense and/or calling the arg `func` rather than `fn` in analogy... probably not, just mentioning.